contracts/Synthswap.sol

```diff
@@ -5,85 +5,295 @@ import "./interfaces/ISynthSwap.sol";
 import "./interfaces/IERC20.sol";
 import "./interfaces/ISynthetix.sol";
 import "./interfaces/IAddressResolver.sol";
+import "./interfaces/IAggregationRouterV4.sol";
+import "./interfaces/IAggregationExecutor.sol";
+import "./utils/SafeERC20.sol";
+import "./libraries/RevertReasonParser.sol";
 
+/// @title system to swap synths to/from many erc20 tokens
+/// @dev IAggregationRouterV4 relies on calldata generated off-chain
 contract SynthSwap is ISynthSwap {
-    ISynthetix synthetix;
-    IERC20 sUSD;
-    IAddressResolver addressResolver;
-    address volumeRewards;
-    address aggregationRouterV3;
+    using SafeERC20 for IERC20;
+
+    IERC20 immutable sUSD;
+    IAggregationRouterV4 immutable router;
+    IAddressResolver immutable addressResolver;
+    address immutable volumeRewards;
+    address immutable treasury;
+
+    address constant ETH_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+    bytes32 private constant CONTRACT_SYNTHETIX = "Synthetix";
+    bytes32 private constant sUSD_CURRENCY_KEY = "sUSD";
+    bytes32 private constant TRACKING_CODE = "KWENTA";
 
     event SwapInto(address indexed from, uint amountReceived);
     event SwapOutOf(address indexed from, uint amountReceived);
+    event Received(address from, uint amountReceived);
     
-    constructor (address _synthetix, address _sUSD, address _volumeRewards, address _aggregationRouterV3, address _addressResolver) {
-        synthetix = ISynthetix(_synthetix);
+    constructor (
+        address _sUSD,
+        address _aggregationRouterV4,
+        address _addressResolver,
+        address _volumeRewards,
+        address _treasury
+    ) {
         sUSD = IERC20(_sUSD);
-        volumeRewards = _volumeRewards;
-        aggregationRouterV3 = _aggregationRouterV3;
+        router = IAggregationRouterV4(_aggregationRouterV4);
         addressResolver = IAddressResolver(_addressResolver);
+        volumeRewards = _volumeRewards;
+        treasury = _treasury;
     }
 
+    //////////////////////////////////////
+    ///////// EXTERNAL FUNCTIONS /////////
+    //////////////////////////////////////
+
+    receive() external payable {
+        emit Received(msg.sender, msg.value);
+    }
+
+    /// @inheritdoc ISynthSwap
     function swapInto(
-        bytes calldata _data,
-        bytes32 destinationSynthCurrencyKey
+        bytes32 _destSynthCurrencyKey,
+        address _destSynthAddress,
```

Comment:
Probably better not the separate these inputs and introduce issues if someone places a mismatching currencyKey & address.

Which function did you use to get a synth on the address resolver? `getAddress` or `getSynth`? https://docs.synthetix.io/contracts/source/contracts/AddressResolver/